src/main/scala/nodes/util/CommonSparseFeatures.scala

```diff
@@ -10,17 +10,31 @@ import scala.reflect.ClassTag
  * An Estimator that chooses the most frequently observed sparse features when training,
  * and produces a transformer which builds a sparse vector out of them
  *
+ * Deterministically orders the feature mappings first by decreasing number of appearances,
+ * then by earliest appearance in the RDD
+ *
  * @param numFeatures The number of features to keep
  */
 case class CommonSparseFeatures[T : ClassTag](numFeatures: Int) extends Estimator[Seq[(T, Double)], SparseVector[Double]] {
   // Ordering that compares (feature, frequency) pairs according to their frequencies
-  val ordering = new Ordering[(T, Int)] {
```

Comment:
zipWithUniqueId luckily doesn't spawn a new job. And yes, newsgroups pipeline has Seq[String] (ngrams) as T, which doesn't have ordering.